phpBB/phpbb/module/module_manager.php

```diff
@@ -0,0 +1,555 @@
+
+				if (!$row)
+				{
+					throw new module_not_found_exception('PARENT_NOT_EXIST');
+				}
+
+				// Workaround
+				$row['left_id'] = (int) $row['left_id'];
+				$row['right_id'] = (int) $row['right_id'];
+
+				$sql = 'UPDATE ' . $this->modules_table . "
+					SET left_id = left_id + 2, right_id = right_id + 2
+					WHERE module_class = '" . $this->db->sql_escape($module_data['module_class']) . "'
+						AND left_id > {$row['right_id']}";
+				$this->db->sql_query($sql);
+
+				$sql = 'UPDATE ' . $this->modules_table . "
+					SET right_id = right_id + 2
+					WHERE module_class = '" . $this->db->sql_escape($module_data['module_class']) . "'
+						AND {$row['left_id']} BETWEEN left_id AND right_id";
+				$this->db->sql_query($sql);
+
+				$module_data['left_id'] = (int) $row['right_id'];
+				$module_data['right_id'] = (int) $row['right_id'] + 1;
+			}
+			else
+			{
+				$sql = 'SELECT MAX(right_id) AS right_id
+					FROM ' . $this->modules_table . "
+					WHERE module_class = '" . $this->db->sql_escape($module_data['module_class']) . "'";
+				$result = $this->db->sql_query($sql);
+				$row = $this->db->sql_fetchrow($result);
+				$this->db->sql_freeresult($result);
+
+				$module_data['left_id'] = (int) $row['right_id'] + 1;
+				$module_data['right_id'] = (int) $row['right_id'] + 2;
+			}
+
+			$sql = 'INSERT INTO ' . $this->modules_table . ' ' . $this->db->sql_build_array('INSERT', $module_data);
+			$this->db->sql_query($sql);
+
+			$module_data['module_id'] = $this->db->sql_nextid();
+		}
+		else
+		{
+			$row = $this->get_module_row($module_data['module_id'], $module_data['module_class']);
+
+			if ($module_data['module_basename'] && !$row['module_basename'])
+			{
+				// we're turning a category into a module
+				$branch = $this->get_module_branch($module_data['module_id'], $module_data['module_class'], 'children', false);
+
+				if (sizeof($branch))
+				{
+					throw new module_not_found_exception('NO_CATEGORY_TO_MODULE');
+				}
+			}
+
+			if ($row['parent_id'] != $module_data['parent_id'])
+			{
+				$this->move_module($module_data['module_id'], $module_data['parent_id'], $module_data['module_class']);
+			}
+
+			$update_ary = $module_data;
+			unset($update_ary['module_id']);
+
+			$sql = 'UPDATE ' . $this->modules_table . '
+				SET ' . $this->db->sql_build_array('UPDATE', $update_ary) . "
+				WHERE module_class = '" . $this->db->sql_escape($module_data['module_class']) . "'
+					AND module_id = " . (int) $module_data['module_id'];
+			$this->db->sql_query($sql);
+		}
+	}
+
+	/**
+	 * Move module around the tree
+	 *
+	 * @param int		$from_module_id	ID of the current parent module
+	 * @param int		$to_parent_id	ID of the target parent module
+	 * @param string	$module_class	Class of the module (acp, ucp, mcp etc...)
+	 */
+	public function move_module($from_module_id, $to_parent_id, $module_class)
+	{
+		$moved_modules = $this->get_module_branch($from_module_id, $module_class, 'children');
```

Comment:
what if there isn't any branch? get_module_branch doesn't throw any exception so the next lines will throw an error?